gen/generate.go

```diff
@@ -115,94 +120,148 @@ func generateGenesisFiles(outDir string, protoVersion protocol.ConsensusVersion,
+					if err != nil {
+						err = fmt.Errorf("could not generate new participation file %s: %v", pfilename, err)
+						os.Remove(pfilename)
+						errorsChannel <- err
+						return
+					}
+					if verbose {
+						verbosedOutput <- fmt.Sprintf("Created new partkey: %s", pfilename)
+					}
+					atomic.AddInt64(&partKeyCreated, 1)
 				}
-				rootKeyCreated++
 			}
 
-			if partkeyErr != nil && wallet.Online == basics.Online {
-				os.Remove(pfilename)
+			var data basics.AccountData
+			data.Status = wallet.Online
+			data.MicroAlgos.Raw = wallet.Stake
+			if wallet.Online == basics.Online {
+				data.VoteID = part.VotingSecrets().OneTimeSignatureVerifier
+				data.SelectionID = part.VRFSecrets().PK
+				data.VoteFirstValid = part.FirstValid
+				data.VoteLastValid = part.LastValid
+				data.VoteKeyDilution = part.KeyDilution
+			}
 
-				partDB, err = db.MakeErasableAccessor(pfilename)
-				if err != nil {
-					err = fmt.Errorf("couldn't open participation DB accessor %s: %v", pfilename, err)
-					os.Remove(pfilename)
-					return
-				}
+			writeMu.Lock()
+			records[wallet.Name] = data
 
-				part, err = account.FillDBWithParticipationKeys(partDB, root.Address(), basics.Round(firstWalletValid), basics.Round(lastWalletValid), partKeyDilution)
-				if err != nil {
-					err = fmt.Errorf("could not generate new participation file %s: %v", pfilename, err)
-					os.Remove(pfilename)
-					return
-				}
-				if verbose {
-					fmt.Printf("Created new partkey: %s\n", pfilename)
-				}
-				partKeyCreated++
+			genesisAddrs[wallet.Name] = root.Address()
+			writeMu.Unlock()
+
+			rootDB.Close()
+			if wallet.Online == basics.Online {
+				partDB.Close()
 			}
 		}
+	}
 
-		var data basics.AccountData
-		data.Status = wallet.Online
-		data.MicroAlgos.Raw = wallet.Stake
-		if wallet.Online == basics.Online {
-			data.VoteID = part.VotingSecrets().OneTimeSignatureVerifier
-			data.SelectionID = part.VRFSecrets().PK
-			data.VoteFirstValid = part.FirstValid
-			data.VoteLastValid = part.LastValid
-			data.VoteKeyDilution = part.KeyDilution
-		}
+	for _, wallet := range allocation {
+		pendingWallets <- wallet
+	}
+
+	if verbose {
+		// create a listener for the verbosedOutput
+		go func() {
+			for textOut := range verbosedOutput {
+				fmt.Printf("%s\n", textOut)
+			}
+		}()
+	}
+
+	creatingWalletsWaitGroup.Add(concurrentWalletGenerators)
+	for routinesCounter := 0; routinesCounter < concurrentWalletGenerators; routinesCounter++ {
+		go createWallet()
+	}
 
-		records[wallet.Name] = data
+	// wait until all goroutines are done.
+	creatingWalletsWaitGroup.Wait()
 
-		genesisAddrs[wallet.Name] = root.Address()
+	close(verbosedOutput)
 
-		rootDB.Close()
-		if wallet.Online == basics.Online {
-			partDB.Close()
-		}
+	// check to see if we had any errors.
+	select {
+	case err := <-errorsChannel:
```

Comment:
read only one error by might have up to concurrentWalletGenerators. It might be possible that one failure cascades others and because of concurrency we read here only secondary error but not the original one.

Reply:
I don't think that there is an elegant solution here, other than reducing concurrency to 1 and trying again.

Reply:
why not reading until it's over? like for + select + break on default.

Reply:
I won't feel bad to have that handled separately by constructing an error object which lists all the errors that were encountered during the process.